mlflow/keras.py

```diff
@@ -461,21 +467,36 @@ def fit(self, *args, **kwargs):
         unlogged_params = ['self', 'x', 'y', 'callbacks', 'validation_data', 'verbose']
 
         log_fn_args_as_params(original, args, kwargs, unlogged_params)
+        early_stop_callback = None
 
         # Checking if the 'callback' argument of fit() is set
         if len(args) >= 6:
             tmp_list = list(args)
+            early_stop_callback = early_stop_check(tmp_list[5])
             tmp_list[5] += [__MLflowKerasCallback()]
             args = tuple(tmp_list)
         elif 'callbacks' in kwargs:
+            early_stop_callback = early_stop_check(kwargs['callbacks'])
             kwargs['callbacks'] += [__MLflowKerasCallback()]
         else:
             kwargs['callbacks'] = [__MLflowKerasCallback()]
 
-        result = original(self, *args, **kwargs)
+        history = original(self, *args, **kwargs)
+
+        if early_stop_callback:
+            try_mlflow_log(mlflow.log_metric, 'stopped_epoch', early_stop_callback.stopped_epoch)
+            if early_stop_callback.restore_best_weights:
+                stopped_metrics = {key: history.history[key][early_stop_callback.stopped_epoch]
+                                   for key in history.history.keys()}
```

Comment:
When the user set `patience` to > 0, the stopped epoch does not correspond to the best epoch.

https://github.com/keras-team/keras/blob/16bd239e34d853f3bf7ae99f14e0324f924ca371/keras/callbacks/callbacks.py#L742

Reply:
Good point, this has to be stopped_epoch - patience, effectively, right?

Reply:
We need to subtract `max(1, patience)`.

Reply:
We also need to consider the case where early stopping didn't occur.

Reply:
Thanks for the catch! Will put the fix in for logging the correct epoch. In the case of early stopping not occurring and `restore_best_weights` is set to be true, the best weights aren't actually restored (see [this issue](https://github.com/keras-team/keras/issues/12511)). Therefore, I don't think we need to find the best epoch when early stopping doesn't happen, though I do need to add a check for not logging an extra step if that is the case.